### PR TITLE
Ignore pip packages in constraints

### PR DESCRIPTION
**Pull request by carlosnewmar**

### CODE REVIEW

File: tests/test_load_yaml_dict.py [hunk 1]
```diff
@@ -250,6 +256,29 @@ def test_constraints_respected(self) -> None:
             "attrs >=20",
         ]
 
+    def test_constraints_respected_with_pip(self) -> None:
```
Comment (nicoddemus) [inline tests/test_load_yaml_dict.py:259]: Nice!

Did you see this test failing, before the fix in `devenv.py` was applied?

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks!

Please also:

* Add a note to https://github.com/ESSS/conda-devenv/blob/master/docs/usage.rst#constraints mentioning that *for now* we are ignoring dependencies declared in `pip` for constraints purposes.
* Add a new CHANGELOG entry.

**Comment by carlosnewmar** [comment]
> Thanks!
> 
> Please also:
> 
> * Add a note to https://github.com/ESSS/conda-devenv/blob/master/docs/usage.rst#constraints mentioning that _for now_ we are ignoring dependencies declared in `pip` for constraints purposes.
> * Add a new CHANGELOG entry.

Thanks for the review Bruno.
I didn't open it to review, because I'll test in pwpa yet.

File: tests/test_load_yaml_dict.py (hunk 1 above)
Comment (carlosnewmar) [reply tests/test_load_yaml_dict.py:259]: Yep. It was failing.
This is the error without the fix:
```
self = <tests.test_load_yaml_dict.TestConstraints object at 0x0000024D8C643290>

    def test_constraints_respected_with_pip(self) -> None:
        """
        Constraints are declared as dependency if they are explicitly declared.
        """
        data = {
            "dependencies": ["attrs >19", "boltons", "pytest>=6", {"pip": ["lupa >= 1.14"]}],
            "constraints": ["pytest >7", "attrs >=20", "requests <2"],
        }
>       process_constraints_into_dependencies(data)

tests\test_load_yaml_dict.py:267:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src\conda_devenv\devenv.py:351: in process_constraints_into_dependencies
    PackageSpecifier.parse(dep).name for dep in yaml_dict["dependencies"] #if isinstance(dep, str)
src\conda_devenv\devenv.py:444: in parse
    m = re.match(cls._PACKAGE_PATTERN, specifier)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

pattern = re.compile('^(?P<channel>[a-z0-9_\\-/.]+::)?(?P<package>[a-z0-9_\\-.]+)\\s*(?P<version>.*)$', re.IGNORECASE), string = {'pip': ['lupa >= 1.14']}, flags = 0

    def match(pattern, string, flags=0):
        """Try to apply the pattern at the start of the string, returning
        a Match object, or None if no match was found."""
>       return _compile(pattern, flags).match(string)
E       TypeError: expected string or bytes-like object, got 'dict'

..\..\envs\devenv\Lib\re\__init__.py:167: TypeError

**Comment by nicoddemus** [comment]
> I didn't open it to review, because I'll test in pwpa yet.

Oh I didn't notice. :sweat_smile:

File: CHANGELOG.rst [hunk 2]
```diff
@@ -2,6 +2,11 @@
 CHANGELOG
 =========
 
+3.2.1 (2024-01-03)
+------------------
+
+* Minor fix to ignore pip constraints and support pip packages in lock files.
```
Comment (nicoddemus) [inline CHANGELOG.rst:8]:
```suggestion
* Fixed handling of `constraints` when processing lock files and `dependencies` include a `pip` section.
```

File: docs/usage.rst [hunk 3]
```diff
@@ -234,6 +234,7 @@ downstream ``devenv.yml`` files will inherit that dependency.
 and `pip <https://pip.pypa.io/en/stable/user_guide/#constraints-files>`__ have the concept of *constraints*,
 which are package version restrictions that do not cause a package to be included in the environment directly,
 but if they are included in the environment by other means, then their version specifiers must be respected.
+However, for the current version dependencies declared in pip are ignored for constraints purposes.
```
Comment (nicoddemus) [inline docs/usage.rst:237]: Let's make this more prominent, moving it next to the `.. note` at the bottom of this section.

```suggestion
```

I suggest to add it as a separate "note", like:



```rst

.. note::
    Currently dependencies via ``pip`` are ignored for constraints purposes.
```

**Review by nicoddemus — APPROVED** [review]
Great work!

**Pull request merged**